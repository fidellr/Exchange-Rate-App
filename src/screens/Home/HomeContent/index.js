import React, { Component } from 'react';
import cls from 'classnames';
import CurrencyInput from 'react-currency-input'
import { convertCurrencyName } from '../../../utils/helpers/convertCurrency';
import RateCard from '../../../components/RateCard';
import AddIcon from '../../../static/icons/add.svg';
import Card from '../../../components/Card';
import Dropdown from '../../../components/Dropdown';
import { isUnique } from '../../../utils/helpers/unique';

class HomeContent extends Component {
  state = {
    currentValue: 10.00,
    selectedRates: null,
    selectedRate: null,
    isDropdownShow: false,
  }

  componentDidMount() {
    const { rates } = this.props;
    const slicedRates = rates.slice(0, 5);
    this.setState({ selectedRates: slicedRates });
  }

  handleChangeCurrency = (value) => {
    if (value.length < 1) {
      this.setState({ currentValue: 10.00 });
      return;
    }
    this.setState({ currentValue: parseFloat(value) });
  }

  toggleShowDropdown = () => {
    const { isDropdownShow } = this.state;
    this.setState({ isDropdownShow: !isDropdownShow });
  }

  handleSubmit = () => {
    const { selectedRates, selectedRate } = this.state;
    if (isUnique(selectedRates, 'rate_name', selectedRate.rate_name)) return;
    this.setState({ selectedRates: [...selectedRates, selectedRate], selectedRate: null });
  }

  handleDropdownChoose = (selectedRate) => this.setState({ selectedRate })

  handleRemove = (rateName) => {
    const { selectedRates } = this.state;
    const updatedRates = selectedRates.filter(item => item.rate_name !== rateName);
    this.setState({ selectedRates: updatedRates });
  }

  renderHeader = () => {
    const { baseCurrency } = this.props
    const { currentValue } = this.state;
    return (
      <React.Fragment>
        <i className="u-fontSize14">{baseCurrency} - {convertCurrencyName(baseCurrency)}</i>
        <div className="u-flex u-justifyContentSpaceBetween u-marginTop6 u-marginBottom0">
          <h3 className="u-margin0">{baseCurrency}</h3>
          <CurrencyInput
            className="currencyInput u-outlineNone u-textAlignRight u-backgroundColorWhite"
            value={currentValue}
            decimalSeparator="."
            thousandSeparator=","
            precision="4"
            onChangeEvent={(e) => this.handleChangeCurrency(e.target.value)}
          />
        </div>
      </React.Fragment>
    )
  }

  renderRateCard = (rates) => {
    const { currentValue } = this.state;
    const { baseCurrency } = this.props;
    return rates && rates.map((item, index) => (
      <div
        key={`rateCard-${item.rate_name}-${Math.random()}-${index}`}
        className={cls('u-border1 u-borderColorGray u-borderRadius2', {
          'u-marginTop16': index !== 0,
        })}
      >
        <RateCard
          onRemove={() => this.handleRemove(item.rate_name)}
          currentBaseValue={currentValue}
          baseCurrency={baseCurrency}
          rateName={item.rate_name}
          rateNumber={item.rate_currency}
        />
      </div>
    ));
  }

  render() {
    const { rates } = this.props;
    const { selectedRates, selectedRate, isDropdownShow } = this.state;

    return (
      <div className="homePage-container u-relative u-paddingBottom10 u-backgroundColorGray">
        <div className="homePage-header u-boxShadowBottom u-zIndex7 u-fixed u-padding16 u-backgroundColorWhite">
          {this.renderHeader()}
        </div>
        <div className="homePage-content u-paddingTop90 u-paddingBottom10 u-paddingHorizontal16">
          {rates && this.renderRateCard(selectedRates)}
        </div>
        <Card className={cls('homePage-bottomInputContainer u-border1 u-borderColorGray u-flex u-justifyContentSpaceBetween u-fontSize14 u-marginHorizontal16 u-backgroundColorWhite', {
          'u-marginBottom16': !isDropdownShow,
        })}>
          <div className="u-cursorPointer u-padding10 u-flex" onClick={this.toggleShowDropdown}>
            <div className="addMore-icon u-backgroundNoRepeat u-backgroundSizeCover u-size24 u-marginAuto" style={{ background: `url(${AddIcon})` }} />
            <div className="addMore-text u-marginAuto u-paddingLeft10">
              {!selectedRate ? `Add More Currencies` : selectedRate.rate_name}
            </div>
          </div>
          {
            selectedRate && (
              <div
                className="u-cursorPointer u-padding14 u-borderLeft u-borderLeftColorGray1 u-sizeFullHeight"
                onClick={this.handleSubmit}
              >
                SUBMIT
            </div>
            )
          }
        </Card>
        {
          isDropdownShow && (
            <Dropdown
              onClick={this.handleDropdownChoose}
              data={rates}
              selectedItems={selectedRates}
            />
          )
        }
      </div>
    )
  }
}

export default HomeContent;