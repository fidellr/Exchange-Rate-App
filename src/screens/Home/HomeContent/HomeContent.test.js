import React from 'react';
import { shallow } from "enzyme";
import HomeContent from './index';

const data = [
  {
    rate_name: 'SGD',
    rate_currency: 11.2213,
  },
  {
    rate_name: 'PLN',
    rate_currency: 22.21134,
  },
  {
    rate_name: 'RUB',
    rate_currency: 33.99823,
  }
];

describe('HomeContent', () => {
  it('should fill selectedRates as the default data that passed from props by its parent', () => {
    const component = shallow(<HomeContent baseCurrency="USD" rates={data} />);
    expect(component.state().selectedRates).toHaveLength(3);
    expect(component).toMatchSnapshot();
  });
});

describe('HomeContent change currency event', () => {
  it('should change currentValue state with the parsed value from react-currency', () => {
    const component = shallow(<HomeContent baseCurrency="USD" rates={data} />);
    expect(component.state().currentValue).toEqual(10.00);
    component.instance().handleChangeCurrency('33.2112');
    expect(component.state().currentValue).toEqual(33.2112);
  });
});

describe('HomeContent show dropdown event', () => {
  it('should set isDropdownShow state to show/ dropdown', () => {
    const component = shallow(<HomeContent baseCurrency="USD" rates={data} />);
    expect(component.state().isDropdownShow).toBeFalsy();
    component.instance().toggleShowDropdown();
    expect(component.state().isDropdownShow).toBeTruthy();
  });
});

describe('HomeContent submit currencyName handler fired', () => {
  it('should merge selectedRates when selectedRate filled after choosing dropdown item', () => {
    const component = shallow(<HomeContent baseCurrency="USD" rates={data} />);
    const selectedRate = { rate_name: 'SEK', rate_currency: 21.3321 };

    component.instance().handleDropdownChoose(selectedRate);
    expect(component.state().selectedRate).toEqual(selectedRate);

    component.instance().handleSubmit();
    expect(component.state().selectedRate).toBeNull();
    expect(component.state().selectedRates).toEqual([...data, selectedRate]);
  });
});

describe('HomeContent remove currency handler', () => {
  it('should remove speicific currency on existing selectedRate state', () => {
    const component = shallow(<HomeContent baseCurrency="USD" rates={data} />);

    component.instance().handleRemove('SGD');
    const mockRemovedPropertyFromSelectedRates = data.filter(item => item.rate_name !== 'SGD');
    expect(component.state().selectedRates).toEqual(mockRemovedPropertyFromSelectedRates);
  });
});
