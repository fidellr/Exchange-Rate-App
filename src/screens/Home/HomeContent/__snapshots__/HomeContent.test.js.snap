// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomeContent should fill selectedRates as the default data that passed from props by its parent 1`] = `
<div
  className="homePage-container u-relative u-paddingBottom10 u-backgroundColorGray"
>
  <div
    className="homePage-header u-boxShadowBottom u-zIndex7 u-fixed u-padding16 u-backgroundColorWhite"
  >
    <i
      className="u-fontSize14"
    >
      USD
       - 
      United States Dollar
    </i>
    <div
      className="u-flex u-justifyContentSpaceBetween u-marginTop6 u-marginBottom0"
    >
      <h3
        className="u-margin0"
      >
        USD
      </h3>
      <CurrencyInput
        allowNegative={false}
        autoFocus={false}
        className="currencyInput u-outlineNone u-textAlignRight u-backgroundColorWhite"
        decimalSeparator="."
        inputType="text"
        onChange={[Function]}
        onChangeEvent={[Function]}
        precision="4"
        prefix=""
        selectAllOnFocus={false}
        suffix=""
        thousandSeparator=","
        value={10}
      />
    </div>
  </div>
  <div
    className="homePage-content u-paddingTop90 u-paddingBottom10 u-paddingHorizontal16"
  >
    <div
      className="u-border1 u-borderColorGray u-borderRadius2"
      key="rateCard-SGD-0.8786105735941141-0"
    >
      <_default
        baseCurrency="USD"
        currentBaseValue={10}
        onRemove={[Function]}
        rateName="SGD"
        rateNumber={11.2213}
      />
    </div>
    <div
      className="u-border1 u-borderColorGray u-borderRadius2 u-marginTop16"
      key="rateCard-PLN-0.9482590948689531-1"
    >
      <_default
        baseCurrency="USD"
        currentBaseValue={10}
        onRemove={[Function]}
        rateName="PLN"
        rateNumber={22.21134}
      />
    </div>
    <div
      className="u-border1 u-borderColorGray u-borderRadius2 u-marginTop16"
      key="rateCard-RUB-0.522197074232398-2"
    >
      <_default
        baseCurrency="USD"
        currentBaseValue={10}
        onRemove={[Function]}
        rateName="RUB"
        rateNumber={33.99823}
      />
    </div>
  </div>
  <_default
    className="homePage-bottomInputContainer u-border1 u-borderColorGray u-flex u-justifyContentSpaceBetween u-fontSize14 u-marginHorizontal16 u-backgroundColorWhite u-marginBottom16"
  >
    <div
      className="u-cursorPointer u-padding10 u-flex"
      onClick={[Function]}
    >
      <div
        className="addMore-icon u-backgroundNoRepeat u-backgroundSizeCover u-size24 u-marginAuto"
        style={
          Object {
            "background": "url(add.svg)",
          }
        }
      />
      <div
        className="addMore-text u-marginAuto u-paddingLeft10"
      >
        Add More Currencies
      </div>
    </div>
  </_default>
</div>
`;
